apps/system/js/sync_state_machine.js

```diff
@@ -0,0 +1,147 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*
+
+       +---------------Success ---------------+
+       |                                      |
+   +---+----+                                 |
+   |Enabling+---Error---+                     |
+   +---+----+           |                     |
+       ^              +-v---+                 |
+       |   +-Disable--+Error+-------------+   |
+       |   |          +--+--+             |   |
+     Enable|             ^             Enable |
+       |   |             |                |   |
+       |   |             |                |   |
+  +----+---v-+           |            +---v---v+
+  | Disabled |         Error +------->+Enabled +---+
+  +----^-----+           |   |        +---+----+   |
+       |                 |   |            |        |
+       |                 | Success       Sync      |
+       |                 |   |            |        |
+       |             +---+---+            |        |
+       |             |Syncing<------------+        |
+       |             +-------+                     |
+       |                                           |
+       +--------------Disable  --------------------+
+
+*/
+
+'use strict';
+
+(function(exports) {
+
+  var SyncStateMachine = {
+    current: 'disabled'
+  };
```

Comment:
Why an object and not a constructor? Would that not give problems if more people want to use a second SyncStateMachine with the same scope?